qcodes/logger/instrument_logger.py

```diff
@@ -54,27 +59,28 @@ class InstrumentFilter(logging.Filter):
     properties as specified in the `extra` dictionary which is a property of
     the adapter.
 
-    Here the ``instrument`` property gets used to reject records that don't
+    Here the ``instrument_name`` property gets used to reject records that don't
     originate from the list of instruments that has been passed to the
     ``__init__`` method.
     """
     def __init__(self, instruments: Union['InstrumentBase',
                                           Sequence['InstrumentBase']]):
-        # This local import is necessary to avoid a circular import dependency.
-        # The alternative is to merge this module with the instrument.base,
-        # which is also not favorable.
         super().__init__()
         if not isinstance(instruments, collections.abc.Sequence):
-            instrument_seq: Sequence['InstrumentBase'] = (instruments,)
+            instrument_seq: Sequence["str"] = (instruments.full_name,)
         else:
-            instrument_seq = instruments
+            instrument_seq = [inst.full_name for inst in instruments]
         self.instrument_set = set(instrument_seq)
 
     def filter(self, record: logging.LogRecord) -> bool:
-        inst: Optional["InstrumentBase"] = getattr(record, "instrument", None)
+        inst: Optional["str"] = getattr(record, "instrument_name", None)
         if inst is None:
             return False
-        return not self.instrument_set.isdisjoint(inst.ancestors)
+
+        insrument_match = any(
+            inst.startswith(instrument_name) for instrument_name in self.instrument_set
```

Comment:
All instrument names are by design unique, so this is safe. Furthermore, the current implementation is inherently broken so that you cannot rely on as an alternative. Do you have any other suggestion. The extra fields must be the primitive types supported by opentelemetry

Reply:
> Do you have any other suggestion

sadly, nope. thankfully we have the naming design the way it is))

> The extra fields must be the primitive types supported by opentelemetry

true. and it's likely not good to add non primitive types as extras to log messages. However, the logging system does not really prevent that, so perhaps there is or there will be a setting on the open telemetry log handler side where we could tell it to ignore some of the keys in the extras.

Reply:
Well that is what happens. Opentelemetry filters that attribute and discards this one with a warning. I still think its a bad idea to embed an object like this. We end up with surprising effects such as the instruments not being garbage collected in tests since there is a reference to it stored in the logs that pytest captures.

Reply:
> We end up with surprising effects such as the instruments not being garbage collected in tests

that's fair. agree.